### PR TITLE
dev-util/sccache: Add 0.3.0

### PR DESCRIPTION
**Pull request by orbea**
Newer rust versions (`1.64.0`) need a newer sccache or there are runtime issues when building rust code, but updating sccache also introduces a new issue for me.

https://github.com/mozilla/sccache/issues/837

In short the problem is that sccache remembers the `${T}` variable from the previous build which fails when it should use the `${T}` from the current  build.

For example if you use sccache to build `dev-util/cargo-ebuild` and then `dev-util/sccache` it will fail.
```
sccache-0.3.0/target/release/build/syn-1f024f804478b580 -L dependency=/var/tmp/portage/dev-util/sccache-0.3.0/work/sccache-0.3.0/target/release/deps --cap-lints allow`
sccache: encountered fatal error
sccache: error: Failed to create temp dir
sccache: caused by: Failed to create temp dir
sccache: caused by: No such file or directory (os error 2) at path "/var/tmp/portage/dev-util/cargo-ebuild-0.5.2/temp/sccacheaxtW5u"
error: could not compile `unicode-xid`
```
This can be worked around fairly easily in `/etc/portage/bashrc` by stopping the sccache server if its currently running in `pre_pkg_setup()`.
```
if [[ "${EBUILD_PHASE}" = "setup" ]]; then

function pre_pkg_setup() {
	# https://github.com/mozilla/sccache/issues/837
	if pidof sccache &> /dev/null; then
		ebegin "pkg_setup hook: sccache --stop-server"
		sccache --stop-server &> /dev/null
		eend $?
	fi
}

fi
```
I'm not sure how to resolve this properly, but this seems to be a case where the problems introduced by updating are less than the current problems when using the older version...

### CODE REVIEW

**Comment by gyakovlev** [comment]
so you are saying that a value from previously built crate for other project gets cached and seeps into future built project that use that cache hit? 

also @etc404 was working on bump too, he has some changes for it I believe.

**Comment by gyakovlev** [comment]
or does it only happen within lifetime of a process and does not actually pollute on-disk cache?

**Comment by etc404** [comment]
I don't believe I have anything to add aside from adding native-zlib to myfeatures in src_configure(), I was still trying to figure out what use flag you were referring to when you mentioned simple-s3 as opposed to s3.

File: dev-util/sccache/sccache-0.3.0.ebuild [hunk 1]
```diff
@@ -0,0 +1,394 @@
...
+	if [[ "${PV}" == *9999* ]]; then
+		git-r3_src_unpack
+		cargo_live_src_unpack
+	else
+		cargo_src_unpack
+	fi
+}
+
+src_configure() {
+	myfeatures=(
```
Comment (gyakovlev) [inline dev-util/sccache/sccache-0.3.0.ebuild:353]: add `native-zlib` here, unconditionally.

File: dev-util/sccache/sccache-0.3.0.ebuild [hunk 2]
```diff
@@ -0,0 +1,394 @@
...
+		$(usev azure)
+		$(usev dist-client)
+		$(usev dist-server)
+		$(usev gcs)
+		$(usev memcached)
+		$(usev redis)
+		$(usev s3)
```
Comment (gyakovlev) [inline dev-util/sccache/sccache-0.3.0.ebuild:360]: it has new feature called `simple-s3` which can be used without `s3`
but, `s3` requires `simple-s3`
so it either needs `REQUIRED_USE`

**Review by gyakovlev — CHANGES_REQUESTED** [review]

File: dev-util/sccache/sccache-0.3.0.ebuild (hunk 2 above)
Comment (etc404) [reply dev-util/sccache/sccache-0.3.0.ebuild:360]: Got it, I'm not sure why pkgcheck was throwing errors about simple-s3 for me.

Comment (gyakovlev) [reply dev-util/sccache/sccache-0.3.0.ebuild:360]: @etc404 you have to add it to metadata.xml in the same directory to satisfy `pkgcheck`

File: dev-util/sccache/sccache-0.3.0.ebuild (hunk 1 above)
Comment (gyakovlev) [reply dev-util/sccache/sccache-0.3.0.ebuild:353]: and ofc don't forget to add a dep

**Comment by orbea** [comment]
> so you are saying that a value from previously built crate for other project gets cached and seeps into future built project that use that cache hit?

The value from the first build will seep into the next future build. This can be easily reproduced on my system by building two different rust projects in a row.

> emerge -av1 sccache cargo-ebuild

File: dev-util/sccache/sccache-0.3.0.ebuild (hunk 2 above)
Comment (orbea) [reply dev-util/sccache/sccache-0.3.0.ebuild:362]: I added the `simple-s3` USE.

File: dev-util/sccache/sccache-0.3.0.ebuild (hunk 1 above)
Comment (orbea) [reply dev-util/sccache/sccache-0.3.0.ebuild:354]: I added `native-zlib`, thanks for the review!

**Comment by gyakovlev** [comment]
> > so you are saying that a value from previously built crate for other project gets cached and seeps into future built project that use that cache hit?
> 
> The value from the first build will seep into the next future build. This can be easily reproduced on my system by building two different rust projects in a row.
> 
> > emerge -av1 sccache cargo-ebuild

but if you restart process it's fine. means it's not the on-disk is corrupt, it's just TOCTOU issue.
it only checks for TMPDIR on startup of the daemon, but does not re-check it again later.

I believe if you'll run `env SCCACHE_NO_DAEMON=1 emerge -av1 sccache cargo-ebuild` it may work fine. can you check it?
sorry I can't install it yet, need access to different machine.

File: dev-util/sccache/sccache-0.3.0.ebuild [hunk 3]
```diff
@@ -0,0 +1,398 @@
...
+
+LICENSE="Apache-2.0 Apache-2.0-with-LLVM-exceptions BSD BSD-2 Boost-1.0 ISC MIT Unlicense ZLIB"
+SLOT="0"
+IUSE="azure dist-client dist-server gcs memcached redis s3 simple-s3"
+REQUIRED_USE="s3? ( simple-s3 )"
+
+BDEPEND="virtual/pkgconfig"
+
+DEPEND="
+	sys-libs/zlib
```
Comment (gyakovlev) [inline dev-util/sccache/sccache-0.3.0.ebuild:333]: needs subslot dep `:=`

File: dev-util/sccache/metadata.xml [hunk 4]
```diff
@@ -11,6 +11,7 @@
 		<flag name="gcs">Enable Google Cloud Storage support</flag>
 		<flag name="redis">Enable Redis support</flag>
 		<flag name="s3">Enable Amazon S3 support</flag>
+		<flag name="simple-s3">Enable Amazon S3 support without s3</flag>
```
Comment (gyakovlev) [inline dev-util/sccache/metadata.xml:14]: description is rather confusing =)
let's do something `Enable minimal Amazon S3 support`
and change other one to `Enable full Amazon S3 support`

**Review by gyakovlev — CHANGES_REQUESTED** [review]
minor changes and good to go.
but if env trick works, we can ship it via `env.d` just for safety.

**Comment by orbea** [comment]
> I believe if you'll run `env SCCACHE_NO_DAEMON=1 emerge -av1 sccache cargo-ebuild` it may work fine. can you check it?

Unfortunately not, with that env the build will hang instead at the start.
```
>>> Compiling source in /var/tmp/portage/dev-util/sccache-0.3.0/work/sccache-0.3.0 ...
 * cargo build --release --features native-zlib --no-default-features
```

File: dev-util/sccache/metadata.xml (hunk 4 above)
Comment (orbea) [reply dev-util/sccache/metadata.xml]: Yes, that is better and I changed it. I wish upstream would of documented the feature so I had a better idea how it differed from the standard `s3` support.

**Comment by gyakovlev** [comment]
maybe also try putting `SCCACHE_IDLE_TIMEOUT=0` to make.conf , like upstream issue suggests.
it's supposed to make server exit right away instead of waiting.
but still will probably not solve if 2 packages building in parallel and use same crate.

anyway, it's merged, thanks for the work.

**Comment by orbea** [comment]
> maybe also try putting `SCCACHE_IDLE_TIMEOUT=0` to make.conf , like upstream issue suggests.

Unfortunately that makes no difference. I was under the impression that causes the server to never time out.

> anyway, it's merged, thanks for the work.

 Thanks!

**Pull request closed without merging**